### PR TITLE
Create magical constructor

### PR DESCRIPTION
**Pull request by kbarnhart**
Addresses Issue #865.

Still to finish/test:
- [x] reading from files
- [x] boundary condition setting
- [x] Synthetic fields in master (PR #827)
- [x] Constructors in master (PR #826)
- [x] Address watershed with field name (PR #870)
- [x] Update to netcdf (PR #873)

@mcflugen might you give me a early-stage review on if there is a better way to do the boundary condition part (that doesn't involve so many if statements). I'd like some way to look up bound functions. For example

```python
bc_params = bc_function_dict[bc_function]
if bc_function in grid.__dir__():
    grid.SOMETHING_HERE[bc_function]( ** bc_params)
```

I've figured out how to use `__dir__` to determine what bound functions are present. I just don't know how to call them yet...

Is this possible?

### CODE REVIEW

**Comment by mcflugen** [comment]
@kbarnhart I'm not totally sure what you're wanting but, if I'm understanding correctly, I think you would be better off using Python's [`inspect`](https://docs.python.org/3.7/library/inspect.html) module.

Try something along the lines of the following,
```python
import inspect
from landlab import RasterModelGrid

grid = RasterModelGrid((3, 4))
methods = dict(inspect.getmembers(grid, inspect.ismethod))
methods[bc_function](*bc_args, **bc_kwds)
```
Do you think that would work for you?

**Comment by kbarnhart** [comment]
@mcflugen I think that is what I want! 

Is there a way to easily provide args and kwargs in a yaml. The only thing I can think of would be something like


```yaml
function_name:
  args:
    - arg1
    - arg2
  kwargs:
    kwarg1: kwarg1
    kwarg2: kwarg2
```

But this adds ANOTHER level to my already deep yaml file.

**Comment by mcflugen** [comment]
@kbarnhart I think that's how you would do it. Maybe you could write it a little more compactly as,
```yaml
function_name: [arg1, arg2, {kwarg1: kwarg1, kwarg2: kwarg2}]
```
It's not exactly the same and would require a little more parsing to separate the *args* from the *kwds* but it could work.

Or, what you had but on a single line,
```yaml
function_name: {args: [arg1, arg2], kwds: {kwarg1: kwarg1, kwarg2: kwarg2}}
```

**Comment by kbarnhart** [comment]
@mcflugen I like the first option:

```yaml
function_name: [arg1, arg2, {kwarg1: kwarg1, kwarg2: kwarg2}]
```

And it wouldn't be hard to parse 
if last element of list is a dictionary, pop it and pass as keyword arguments.

I think it makes sense to support this. Do you? Alternatively everything gets passed as keyword arguments.

**Comment by kbarnhart** [comment]
Docs for this are in progress. Eventually they will be built here: https://landlab.readthedocs.io/en/barnhark-magical_constructor/

**Comment by kbarnhart** [comment]
@mcflugen this is now ready for review. 

Docs are [here](https://landlab.readthedocs.io/en/barnhark-magical_constructor/landlab.grid.create.html)

In addition to the new constructor, it adds a few things to the read functions of esri ascii and netcdf to do grid compatibility checking. 

Appveyor tests are failing on the docstring for `Landlab.core.load_params.load_params`. 

Not sure why this is, as I didn't change this file...

My assumption is that PRs #873 #870 and #827 will be merged before this one is. 

```
   def load_params(file_like):
3085        """Load parameters from a file.
3086    
3087        Parameters
3088        ----------
3089        file_like : file_like or str
3090            Contents of a parameter file, a file-like object, or the path to
3091            a parameter file.
3092    
3093        Returns
3094        -------
3095        dict
3096            Parameters as key-value pairs.
3097    
3098        Examples
3099        --------
3100        >>> from landlab.core import load_params
3101        >>> contents = \"\"\"
3102        ... start: 0.
3103        ... stop: 10.
3104        ... step: 2.
3105        ... \"\"\"
3106        >>> params = load_params(contents)
3107        >>> isinstance(params, dict)
3108        True
3109        >>> params['start'], params['stop'], params['step']
3110        (0.0, 10.0, 2.0)
3111    
3112        >>> contents = \"\"\"
3113        ... start: Start time
3114        ... 0.
3115        ... stop: Stop time
3116        ... 10.
3117        ... step: Step time
3118        ... 2.
3119        ... \"\"\"
3120        >>> params = load_params(contents)
3121        >>> isinstance(params, dict)
3122        True
3123        >>> params['start'], params['stop'], params['step']
3124        (0.0, 10.0, 2.0)
3125        """
3126        contents = load_file_contents(file_like)
3127    
3128        try:
3129            params = yaml.load(contents, Loader=_loader)
3130        except yaml.YAMLError:
3131            file_like = six.StringIO(contents)
3132            params = ModelParameterDictionary(from_file=file_like, auto_type=True)
3133    
3134        if not isinstance(params, dict):
3135>           raise ValueError("parsing of parameter file did not produce a dict-like object")
3136E           ValueError: parsing of parameter file did not produce a dict-like object
```

File: landlab/grid/create.py [hunk 1]
```diff
@@ -84,3 +111,252 @@ def create_and_initialize_grid(input_source):
...
+    either be created by reading from a file or creating synthetic values. The
+    :py:func:`~landlab.io.netcdf.read.read_netcdf` and
+    :py:func:`~landlab.io.esri_ascii.read_esri_ascii` functions, and the
+    :py:mod:`synthetic fields <landlab.values.synthetic>`
+    package are currently supported methods to create fields. These may be
+    chained together (as is shown in the Example section below). If these
+    functions do not meet your needs, we welcome contributions that extend the
+    capabilities of this function.
+
+    The following example would uses the
```
Comment (gregtucker) [inline landlab/grid/create.py]: Add text explaining the usage of plane and normal

File: landlab/grid/create.py [hunk 2]
```diff
@@ -84,3 +111,252 @@ def create_and_initialize_grid(input_source):
 
     # Return the created and initialized grid
     return grid_reader(param_dict)
+
+
+def create_grid(file_like):
+    """Create grid, initialize fields, and set boundary conditions.
+
+    **create_grid** expects a dictionary with three keys "grid", "fields", and
```
Comment (gregtucker) [inline landlab/grid/create.py]:
```suggestion
    **create_grid** expects a dictionary with three keys: "grid", "fields", and
```

File: landlab/grid/create.py (hunk 1 above)
Comment (kbarnhart) [reply landlab/grid/create.py]: Done! Good idea!

File: landlab/grid/create.py (hunk 2 above)
Comment (kbarnhart) [reply landlab/grid/create.py]: Done!

**Comment by kbarnhart** [comment]
@mcflugen I've dug into the failing test. I think it is because `simple_create.yaml` isn't being parsed correctly. I'll work on fixing this.

**Comment by mcflugen** [comment]
👍 Looks good @kbarnhart

**Comment by kbarnhart** [comment]
@mcflugen, assuming this passes, ok with you if I make a new release? 1.6.1? 1.7? I think 1.7 is more appropriate.

**Comment by mcflugen** [comment]
@kbarnhart ok with me. 1.7 sounds good.
